typhos/utils.py

```diff
@@ -517,3 +519,200 @@ def find_file_in_paths(filename, *, paths=None):
                 yield match
+
+
+@contextlib.contextmanager
+def subscription_context(*objects, callback, event_type=None, run=True):
+    '''
+    [Context manager] Subscribe to a specific event from all objects
+
+    Unsubscribes all signals before exiting
+
+    Parameters
+    ----------
+    *objects : ophyd.OphydObj
+        Ophyd objects (signals) to monitor
+    callback : callable
+        Callback to run, with same signature as that of
+        :meth:``ophyd.OphydObj.subscribe``
+    event_type : str, optional
+        The event type to subscribe to
+    run : bool, optional
+        Run the previously cached subscription immediately
+    '''
+    obj_to_cid = {}
+    try:
+        for obj in objects:
+            obj_to_cid[obj] = obj.subscribe(callback, event_type=event_type,
+                                            run=run)
+        yield dict(obj_to_cid)
+    finally:
+        for obj, cid in obj_to_cid.items():
+            obj.unsubscribe(cid)
+
+
+def get_all_signals_from_device(device, include_lazy=False, filter_by=None):
+    '''
+    Get all signals in a given device
+
+    Parameters
+    ----------
+    device : ophyd.Device
+        ophyd Device to monitor
+    include_lazy : bool, optional
+        Include lazy signals as well
+    filter_by : callable, optional
+        Filter signals, with signature ``callable(ophyd.Device.ComponentWalk)``
+    '''
+    if not filter_by:
+        def filter_by(walk):
+            return True
+
+    def _get_signals():
+        return [
+            walk.item
+            for walk in device.walk_signals(include_lazy=include_lazy)
+            if filter_by(walk)
+        ]
+
+    if not include_lazy:
+        return _get_signals()
+
+    with no_device_lazy_load():
+        return _get_signals()
+
+
+@contextlib.contextmanager
+def subscription_context_device(device, callback, event_type=None, run=True, *,
+                                include_lazy=False, filter_by=None):
+    '''
+    [Context manager] Subscribe to ``event_type`` from signals in ``device``
+
+    Unsubscribes all signals before exiting
+
+    Parameters
+    ----------
+    device : ophyd.Device
+        ophyd Device to monitor
+    callback : callable
+        Callback to run, with same signature as that of
+        :meth:``ophyd.OphydObj.subscribe``
+    event_type : str, optional
+        The event type to subscribe to
+    run : bool, optional
+        Run the previously cached subscription immediately
+    include_lazy : bool, optional
+        Include lazy signals as well
+    filter_by : callable, optional
+        Filter signals, with signature ``callable(ophyd.Device.ComponentWalk)``
+    '''
+    signals = get_all_signals_from_device(device, include_lazy=include_lazy)
+    with subscription_context(*signals, callback=callback,
+                              event_type=event_type, run=run) as obj_to_cid:
+        yield obj_to_cid
+
+
+class _ConnectionStatus:
+    def __init__(self, signals, callback):
```

Comment:
`signals` is an unused argument here

Reply:
Good catch - fixed in 35c850d. 

Do you think the class itself adds any value? That is, it's yielded by the context manager and summarizes connection status. It could be up to the caller to keep track of it.

Reply:
I can't imagine any specific use case, but it seems useful in abstract

Reply:
OK 👍 

Hmm, maybe it could gain a `remove_signal` method in the future, or something along those lines. We'll see how it's used and go from there.